File: Python/P0002.py
from typing import Generator


def fibonacci(value: int) -> Generator[int, None, None]:
    sequence = {0: 1, 1: 1}
    count = 0
    while sequence[list(sequence)[-1]] < value:
        if count not in sequence:
            sequence[count] = sequence[count - 1] + sequence[count - 2]
        yield sequence[count]
        count += 1


def is_even(number: int) -> bool:
    return number & 1 == 0


def main() -> None:
    answer = sum(filter(is_even, fibonacci(4000000)))
    print(f"Answer: {answer}")


if __name__ == '__main__':
    main()

File: Go/P0002.go
package main

import (
	"fmt"
)

func is_even(number int) bool {
	return number%2 == 0
}

func fibonacci(threshold int) chan int {
	sequence := make(chan int)
	go func() {
		x, y := 1, 1
		for x <= threshold {
			sequence <- x
			x, y = y, x+y
		}
		close(sequence)
	}()
	return sequence
}

func calculate_sum(threshold int) int {
	var sum int = 0
	for number := range fibonacci(threshold) {
		if is_even(number) {
			sum += number
		}
	}
	return sum
}

func main() {
	var threshold int = 4000000
	var answer int = calculate_sum(threshold)
	fmt.Printf("Answer: %d\n", answer)
}
